Pybind fails to build with MSVC 19.16 in C++17 mode

## Fails to build with MSVC 19.16.27034.0 in C++17
I used pybind v 2.4.3

> cmake -G Ninja -DPYBIND11_CPP_STANDARD=/std:c++17 ..\src
> ninja

Error:
> D:\Libs\Pybind\2.4.3\src\include\pybind11/stl_bind.h(415): error C2893: Failed to specialize function template 'enable_if<_Test,_Ty>::type pybind11::detail::vector_buffer(Class_ &)'
        with
        [
            _Ty=void
        ]

No problem with the default c++14 mode.

Great library, BTW! 😃 👍

@YannickJadoul , huh, strange!

I re-tested it again, and here's the full log:

## 64 bit Windows, cmake and ninja used from VS 2017, pybind 2.5.0
```
(pybind11_test) D:\Libs\Pybind\2.5.0\build_ninja>where cmake
C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe
(pybind11_test) D:\Libs\Pybind\2.5.0\build_ninja>where ninja
C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\Common7\IDE\CommonExtensions\Microsoft\CMake\Ninja\ninja.exe

(pybind11_test) D:\Libs\Pybind\2.5.0\build_ninja>cmake --version
cmake version 3.12.18081601-MSVC_2

CMake suite maintained and supported by Kitware (kitware.com/cmake).
(pybind11_test) D:\Libs\Pybind\2.5.0\build_ninja>ninja --version
1.8.2
```
```
(pybind11_test) D:\Libs\Pybind\2.5.0\build_ninja>cls && cmake -G Ninja -DPYBIND11_CPP_STANDARD=/std:c++17 ..\src
-- The CXX compiler identification is MSVC 19.16.27038.0
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Professional/VC/Tools/MSVC/14.16.27023/bin/Hostx64/x64/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Professional/VC/Tools/MSVC/14.16.27023/bin/Hostx64/x64/cl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PythonInterp: D:/Programs/Miniconda3/envs/pybind11_test/python.exe (found version "3.6.10")
-- Found PythonLibs: D:/Programs/Miniconda3/envs/pybind11_test/libs/Python36.lib
-- Building tests WITHOUT Eigen
-- Could NOT find Boost
-- Performing Test HAS_MSVC_GL_LTCG
-- Performing Test HAS_MSVC_GL_LTCG - Success
-- LTO enabled
-- Catch not detected. Interpreter tests will be skipped. Install Catch headers manually or use `cmake -DDOWNLOAD_CATCH=1` to fetch them automatically.
-- pybind11 v2.5.0
-- Configuring done
-- Generating done
-- Build files have been written to: D:/Libs/Pybind/2.5.0/build_ninja

(pybind11_test) D:\Libs\Pybind\2.5.0\build_ninja>cmake --build .
[1/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_async.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[2/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_chrono.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[3/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_call_policies.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[4/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_constants_and_functions.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[5/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\pybind11_tests.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[6/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_callbacks.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[7/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_buffers.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[8/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_builtin_casters.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[9/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_copy_move.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[10/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_class.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[11/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_exceptions.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[12/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_docstring_options.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[13/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_enum.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[14/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_eval.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[15/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_iostream.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[16/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_gil_scoped.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[17/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_local_bindings.cpp.obj
FAILED: tests/CMakeFiles/pybind11_tests.dir/test_local_bindings.cpp.obj
C:\PROGRA~2\MICROS~2\2017\PROFES~1\VC\Tools\MSVC\1416~1.270\bin\Hostx64\x64\cl.exe  /nologo /TP -Dpybind11_tests_EXPORTS -ID:\Libs\Pybind\2.5.0\src\include -ID:\Programs\Miniconda3\envs\pybind11_test\include /DWIN32 /D_WINDOWS /W3 /GR /EHsc /MDd /Zi /Ob0 /Od /RTC1   /std:c++17 /bigobj /W4 /utf-8 /showIncludes /Fotests\CMakeFiles\pybind11_tests.dir\test_local_bindings.cpp.obj /Fdtests\CMakeFiles\pybind11_tests.dir\ /FS -c D:\Libs\Pybind\2.5.0\src\tests\test_local_bindings.cpp
cl : Command line warning D9025 : overriding '/W3' with '/W4'
D:\Libs\Pybind\2.5.0\src\include\pybind11/stl_bind.h(434): error C2672: 'pybind11::detail::vector_buffer': no matching overloaded function found
D:\Libs\Pybind\2.5.0\src\tests\test_local_bindings.cpp(53): note: see reference to function template instantiation 'pybind11::class_<LocalVec,std::unique_ptr<Vector,std::default_delete<_Ty>>> pybind11::bind_vector<LocalVec,std::unique_ptr<_Ty,std::default_delete<_Ty>>,>(pybind11::handle,const std::string &)' being compiled
        with
        [
            Vector=LocalVec,
            _Ty=LocalVec
        ]
d:\libs\pybind\2.5.0\src\tests\pybind11_tests.h(60): note: see reference to class template instantiation 'pybind11::detail::descr<12>' being compiled
d:\libs\pybind\2.5.0\src\include\pybind11\cast.h(1602): note: see reference to class template instantiation 'pybind11::detail::descr<8>' being compiled
d:\libs\pybind\2.5.0\src\include\pybind11\cast.h(1600): note: see reference to class template instantiation 'pybind11::detail::descr<5>' being compiled
d:\libs\pybind\2.5.0\src\include\pybind11\cast.h(1145): note: see reference to class template instantiation 'pybind11::detail::descr<7>' being compiled
D:\Libs\Pybind\2.5.0\src\include\pybind11/stl_bind.h(422): error C2893: Failed to specialize function template 'enable_if<_Test,_Ty>::type pybind11::detail::vector_buffer(Class_ &)'
        with
        [
            _Ty=void
        ]
D:\Libs\Pybind\2.5.0\src\include\pybind11/stl_bind.h(422): note: With the following template arguments:
D:\Libs\Pybind\2.5.0\src\include\pybind11/stl_bind.h(422): note: 'Vector=Vector'
D:\Libs\Pybind\2.5.0\src\include\pybind11/stl_bind.h(422): note: 'Class_=Class_'
D:\Libs\Pybind\2.5.0\src\include\pybind11/stl_bind.h(422): note: 'Args={}'
[18/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_kwargs_and_defaults.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[19/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_modules.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[20/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_factory_constructors.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[21/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_numpy_vectorize.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[22/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_numpy_dtypes.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[23/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_multiple_inheritance.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[24/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_methods_and_attributes.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[25/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_numpy_array.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
[26/41] Building CXX object tests\CMakeFiles\pybind11_tests.dir\test_sequences_and_iterators.cpp.obj
cl : Command line warning D9025 : overriding '/W3' with '/W4'
ninja: build stopped: subcommand failed.
```
Exactly the same error with **python 3.6.10, 3.7.6, 3.8.3.**

I tried with "Visual Studio 15 2017 Win64" generator, too, but it produced the same error:

## 64 bit Windows, pybind 2.5.0, python 3.6.10
```
(pybind11_test) D:\Libs\Pybind\2.5.0\build_ninja>cls && cmake -G "Visual Studio 15 2017 Win64" -DPYBIND11_CPP_STANDARD=/std:c++17 ..\src
-- Selecting Windows SDK version 10.0.17763.0 to target Windows 10.0.18363.
-- The CXX compiler identification is MSVC 19.16.27038.0
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Professional/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Professional/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/cl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PythonInterp: D:/Programs/Miniconda3/envs/pybind11_test/python.exe (found version "3.6.10")
-- Found PythonLibs: D:/Programs/Miniconda3/envs/pybind11_test/libs/Python36.lib
-- Building tests WITHOUT Eigen
-- Could NOT find Boost
-- Performing Test HAS_MSVC_GL_LTCG
-- Performing Test HAS_MSVC_GL_LTCG - Success
-- LTO enabled
-- Catch not detected. Interpreter tests will be skipped. Install Catch headers manually or use `cmake -DDOWNLOAD_CATCH=1` to fetch them automatically.
-- pybind11 v2.5.0
-- Configuring done
-- Generating done
-- Build files have been written to: D:/Libs/Pybind/2.5.0/build_ninja

(pybind11_test) D:\Libs\Pybind\2.5.0\build_ninja>cmake --build .
Microsoft (R) Build Engine version 15.9.21+g9802d43bc3 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 7/10/2020 6:27:16 PM.
Project "D:\Libs\Pybind\2.5.0\build_ninja\ALL_BUILD.vcxproj" on node 1 (default targets).
Project "D:\Libs\Pybind\2.5.0\build_ninja\ALL_BUILD.vcxproj" (1) is building "D:\Libs\Pybind\2.5.0\build_ninja\ZERO_CHE
CK.vcxproj" (2) on node 1 (default targets).
PrepareForBuild:
  Creating directory "x64\Debug\ZERO_CHECK\".
  Creating directory "x64\Debug\ZERO_CHECK\ZERO_CHECK.tlog\".
InitializeBuildStatus:
  Creating "x64\Debug\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
CustomBuild:
  Checking Build System
  CMake does not need to re-run because D:/Libs/Pybind/2.5.0/build_ninja/CMakeFiles/generate.stamp is up-to-date.
  CMake does not need to re-run because D:/Libs/Pybind/2.5.0/build_ninja/tests/CMakeFiles/generate.stamp is up-to-date.
  CMake does not need to re-run because D:/Libs/Pybind/2.5.0/build_ninja/tests/test_embed/CMakeFiles/generate.stamp is
  up-to-date.
  CMake does not need to re-run because D:/Libs/Pybind/2.5.0/build_ninja/tests/test_cmake_build/CMakeFiles/generate.sta
  mp is up-to-date.
FinalizeBuildStatus:
  Deleting file "x64\Debug\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild".
  Touching "x64\Debug\ZERO_CHECK\ZERO_CHECK.tlog\ZERO_CHECK.lastbuildstate".
Done Building Project "D:\Libs\Pybind\2.5.0\build_ninja\ZERO_CHECK.vcxproj" (default targets).

Project "D:\Libs\Pybind\2.5.0\build_ninja\ALL_BUILD.vcxproj" (1) is building "D:\Libs\Pybind\2.5.0\build_ninja\tests\cr
oss_module_gil_utils.vcxproj" (3) on node 1 (default targets).
PrepareForBuild:
  Creating directory "cross_module_gil_utils.dir\Debug\".
  Creating directory "cross_module_gil_utils.dir\Debug\cross_mo.6D4D950E.tlog\".
InitializeBuildStatus:
  Creating "cross_module_gil_utils.dir\Debug\cross_mo.6D4D950E.tlog\unsuccessfulbuild" because "AlwaysCreate" was speci
  fied.
CustomBuild:
  Building Custom Rule D:/Libs/Pybind/2.5.0/src/tests/CMakeLists.txt
  CMake does not need to re-run because D:/Libs/Pybind/2.5.0/build_ninja/tests/CMakeFiles/generate.stamp is up-to-date.
ClCompile:
  D:\Tools\BullseyeCoverage\bin\CL.exe /c /ID:\Libs\Pybind\2.5.0\src\include /ID:\Programs\Miniconda3\envs\pybind11_tes
  t\include /I"C:\Programs\Visual Leak Detector\include" /Zi /nologo /W4 /WX- /diagnostics:classic /Od /Ob0 /D WIN32 /D
   _WINDOWS /D "CMAKE_INTDIR=\"Debug\"" /D cross_module_gil_utils_EXPORTS /D _WINDLL /D _MBCS /Gm- /EHsc /RTC1 /MDd /GS
   /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /GR /std:c++17 /Fo"cross_module_gil_utils.dir\Debug\\" /Fd"cross_mod
  ule_gil_utils.dir\Debug\vc141.pdb" /Gd /TP /FC /errorReport:queue  /bigobj /utf-8 D:\Libs\Pybind\2.5.0\src\tests\cros
  s_module_gil_utils.cpp
  cross_module_gil_utils.cpp
MakeDirsForLink:
  Creating directory "D:\Libs\Pybind\2.5.0\build_ninja\tests\Debug\".
Link:
  D:\Tools\BullseyeCoverage\bin\link.exe /ERRORREPORT:QUEUE /OUT:"D:\Libs\Pybind\2.5.0\src\tests\cross_module_gil_utils
  .cp36-win_amd64.pyd" /INCREMENTAL /NOLOGO /LIBPATH:"C:\Programs\Visual Leak Detector\lib\Win64" D:\Programs\Miniconda
  3\envs\pybind11_test\libs\Python36.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.
  lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /D
  EBUG /PDB:"D:/Libs/Pybind/2.5.0/src/tests/cross_module_gil_utils.pdb" /SUBSYSTEM:CONSOLE /TLBID:1 /DYNAMICBASE /NXCOM
  PAT /IMPLIB:"D:/Libs/Pybind/2.5.0/build_ninja/tests/Debug/cross_module_gil_utils.lib" /MACHINE:X64  /machine:x64 /DLL
   cross_module_gil_utils.dir\Debug\cross_module_gil_utils.obj
     Creating library D:/Libs/Pybind/2.5.0/build_ninja/tests/Debug/cross_module_gil_utils.lib and object D:/Libs/Pybind
  /2.5.0/build_ninja/tests/Debug/cross_module_gil_utils.exp
  cross_module_gil_utils.vcxproj -> D:\Libs\Pybind\2.5.0\src\tests\cross_module_gil_utils.cp36-win_amd64.pyd
FinalizeBuildStatus:
  Deleting file "cross_module_gil_utils.dir\Debug\cross_mo.6D4D950E.tlog\unsuccessfulbuild".
  Touching "cross_module_gil_utils.dir\Debug\cross_mo.6D4D950E.tlog\cross_module_gil_utils.lastbuildstate".
Done Building Project "D:\Libs\Pybind\2.5.0\build_ninja\tests\cross_module_gil_utils.vcxproj" (default targets).

Project "D:\Libs\Pybind\2.5.0\build_ninja\ALL_BUILD.vcxproj" (1) is building "D:\Libs\Pybind\2.5.0\build_ninja\tests\py
bind11_cross_module_tests.vcxproj" (4) on node 1 (default targets).
PrepareForBuild:
  Creating directory "pybind11_cross_module_tests.dir\Debug\".
  Creating directory "pybind11_cross_module_tests.dir\Debug\pybind11.5499A21F.tlog\".
InitializeBuildStatus:
  Creating "pybind11_cross_module_tests.dir\Debug\pybind11.5499A21F.tlog\unsuccessfulbuild" because "AlwaysCreate" was
  specified.
CustomBuild:
  Building Custom Rule D:/Libs/Pybind/2.5.0/src/tests/CMakeLists.txt
  CMake does not need to re-run because D:/Libs/Pybind/2.5.0/build_ninja/tests/CMakeFiles/generate.stamp is up-to-date.
ClCompile:
  D:\Tools\BullseyeCoverage\bin\CL.exe /c /ID:\Libs\Pybind\2.5.0\src\include /ID:\Programs\Miniconda3\envs\pybind11_tes
  t\include /I"C:\Programs\Visual Leak Detector\include" /Zi /nologo /W4 /WX- /diagnostics:classic /Od /Ob0 /D WIN32 /D
   _WINDOWS /D "CMAKE_INTDIR=\"Debug\"" /D pybind11_cross_module_tests_EXPORTS /D _WINDLL /D _MBCS /Gm- /EHsc /RTC1 /MD
  d /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /GR /std:c++17 /Fo"pybind11_cross_module_tests.dir\Debug\\" /Fd
  "pybind11_cross_module_tests.dir\Debug\vc141.pdb" /Gd /TP /FC /errorReport:queue  /bigobj /utf-8 D:\Libs\Pybind\2.5.0
  \src\tests\pybind11_cross_module_tests.cpp
  pybind11_cross_module_tests.cpp
d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(434): error C2672: 'pybind11::detail::vector_buffer': no matching
overloaded function found [D:\Libs\Pybind\2.5.0\build_ninja\tests\pybind11_cross_module_tests.vcxproj]
  d:\libs\pybind\2.5.0\src\tests\pybind11_cross_module_tests.cpp(52): note: see reference to function template instanti
  ation 'pybind11::class_<LocalVec,std::unique_ptr<Vector,std::default_delete<_Ty>>> pybind11::bind_vector<LocalVec,std
  ::unique_ptr<_Ty,std::default_delete<_Ty>>,>(pybind11::handle,const std::string &)' being compiled
          with
          [
              Vector=LocalVec,
              _Ty=LocalVec
          ]
  d:\libs\pybind\2.5.0\src\tests\pybind11_tests.h(60): note: see reference to class template instantiation 'pybind11::d
  etail::descr<12>' being compiled
  d:\libs\pybind\2.5.0\src\include\pybind11\cast.h(1602): note: see reference to class template instantiation 'pybind11
  ::detail::descr<8>' being compiled
  d:\libs\pybind\2.5.0\src\include\pybind11\cast.h(1600): note: see reference to class template instantiation 'pybind11
  ::detail::descr<5>' being compiled
  d:\libs\pybind\2.5.0\src\include\pybind11\cast.h(1145): note: see reference to class template instantiation 'pybind11
  ::detail::descr<7>' being compiled
d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(422): error C2893: Failed to specialize function template 'enable_
if<_Test,_Ty>::type pybind11::detail::vector_buffer(Class_ &)' [D:\Libs\Pybind\2.5.0\build_ninja\tests\pybind11_cross_m
odule_tests.vcxproj]
          with
          [
              _Ty=void
          ]
  d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(422): note: With the following template arguments:
  d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(422): note: 'Vector=Vector'
  d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(422): note: 'Class_=Class_'
  d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(422): note: 'Args={}'
Done Building Project "D:\Libs\Pybind\2.5.0\build_ninja\tests\pybind11_cross_module_tests.vcxproj" (default targets) --
 FAILED.

Project "D:\Libs\Pybind\2.5.0\build_ninja\ALL_BUILD.vcxproj" (1) is building "D:\Libs\Pybind\2.5.0\build_ninja\tests\py
bind11_tests.vcxproj" (5) on node 1 (default targets).
PrepareForBuild:
  Creating directory "pybind11_tests.dir\Debug\".
  Creating directory "pybind11_tests.dir\Debug\pybind11_tests.tlog\".
InitializeBuildStatus:
  Creating "pybind11_tests.dir\Debug\pybind11_tests.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
CustomBuild:
  Building Custom Rule D:/Libs/Pybind/2.5.0/src/tests/CMakeLists.txt
  CMake does not need to re-run because D:/Libs/Pybind/2.5.0/build_ninja/tests/CMakeFiles/generate.stamp is up-to-date.
ClCompile:
  D:\Tools\BullseyeCoverage\bin\CL.exe /c /ID:\Libs\Pybind\2.5.0\src\include /ID:\Programs\Miniconda3\envs\pybind11_tes
  t\include /I"C:\Programs\Visual Leak Detector\include" /Zi /nologo /W4 /WX- /diagnostics:classic /Od /Ob0 /D WIN32 /D
   _WINDOWS /D "CMAKE_INTDIR=\"Debug\"" /D pybind11_tests_EXPORTS /D _WINDLL /D _MBCS /Gm- /EHsc /RTC1 /MDd /GS /fp:pre
  cise /Zc:wchar_t /Zc:forScope /Zc:inline /GR /std:c++17 /Fo"pybind11_tests.dir\Debug\\" /Fd"pybind11_tests.dir\Debug\
  vc141.pdb" /Gd /TP /FC /errorReport:queue  /bigobj /utf-8 D:\Libs\Pybind\2.5.0\src\tests\pybind11_tests.cpp D:\Libs\P
  ybind\2.5.0\src\tests\test_async.cpp D:\Libs\Pybind\2.5.0\src\tests\test_buffers.cpp D:\Libs\Pybind\2.5.0\src\tests\t
  est_builtin_casters.cpp D:\Libs\Pybind\2.5.0\src\tests\test_call_policies.cpp D:\Libs\Pybind\2.5.0\src\tests\test_cal
  lbacks.cpp D:\Libs\Pybind\2.5.0\src\tests\test_chrono.cpp D:\Libs\Pybind\2.5.0\src\tests\test_class.cpp D:\Libs\Pybin
  d\2.5.0\src\tests\test_constants_and_functions.cpp D:\Libs\Pybind\2.5.0\src\tests\test_copy_move.cpp D:\Libs\Pybind\2
  .5.0\src\tests\test_docstring_options.cpp D:\Libs\Pybind\2.5.0\src\tests\test_enum.cpp D:\Libs\Pybind\2.5.0\src\tests
  \test_eval.cpp D:\Libs\Pybind\2.5.0\src\tests\test_exceptions.cpp D:\Libs\Pybind\2.5.0\src\tests\test_factory_constru
  ctors.cpp D:\Libs\Pybind\2.5.0\src\tests\test_gil_scoped.cpp D:\Libs\Pybind\2.5.0\src\tests\test_iostream.cpp D:\Libs
  \Pybind\2.5.0\src\tests\test_kwargs_and_defaults.cpp D:\Libs\Pybind\2.5.0\src\tests\test_local_bindings.cpp D:\Libs\P
  ybind\2.5.0\src\tests\test_methods_and_attributes.cpp D:\Libs\Pybind\2.5.0\src\tests\test_modules.cpp D:\Libs\Pybind\
  2.5.0\src\tests\test_multiple_inheritance.cpp D:\Libs\Pybind\2.5.0\src\tests\test_numpy_array.cpp D:\Libs\Pybind\2.5.
  0\src\tests\test_numpy_dtypes.cpp D:\Libs\Pybind\2.5.0\src\tests\test_numpy_vectorize.cpp D:\Libs\Pybind\2.5.0\src\te
  sts\test_opaque_types.cpp D:\Libs\Pybind\2.5.0\src\tests\test_operator_overloading.cpp D:\Libs\Pybind\2.5.0\src\tests
  \test_pickling.cpp D:\Libs\Pybind\2.5.0\src\tests\test_pytypes.cpp D:\Libs\Pybind\2.5.0\src\tests\test_sequences_and_
  iterators.cpp D:\Libs\Pybind\2.5.0\src\tests\test_smart_ptr.cpp D:\Libs\Pybind\2.5.0\src\tests\test_stl.cpp D:\Libs\P
  ybind\2.5.0\src\tests\test_stl_binders.cpp D:\Libs\Pybind\2.5.0\src\tests\test_tagbased_polymorphic.cpp D:\Libs\Pybin
  d\2.5.0\src\tests\test_union.cpp D:\Libs\Pybind\2.5.0\src\tests\test_virtual_functions.cpp
  pybind11_tests.cpp
  test_async.cpp
  test_buffers.cpp
  test_builtin_casters.cpp
  test_call_policies.cpp
  test_callbacks.cpp
  test_chrono.cpp
  test_class.cpp
  test_constants_and_functions.cpp
  test_copy_move.cpp
  test_docstring_options.cpp
  test_enum.cpp
  test_eval.cpp
  test_exceptions.cpp
  test_factory_constructors.cpp
  test_gil_scoped.cpp
  test_iostream.cpp
  test_kwargs_and_defaults.cpp
  test_local_bindings.cpp
d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(434): error C2672: 'pybind11::detail::vector_buffer': no matching
overloaded function found [D:\Libs\Pybind\2.5.0\build_ninja\tests\pybind11_tests.vcxproj]
  d:\libs\pybind\2.5.0\src\tests\test_local_bindings.cpp(53): note: see reference to function template instantiation 'p
  ybind11::class_<LocalVec,std::unique_ptr<Vector,std::default_delete<_Ty>>> pybind11::bind_vector<LocalVec,std::unique
  _ptr<_Ty,std::default_delete<_Ty>>,>(pybind11::handle,const std::string &)' being compiled
          with
          [
              Vector=LocalVec,
              _Ty=LocalVec
          ]
  d:\libs\pybind\2.5.0\src\tests\pybind11_tests.h(60): note: see reference to class template instantiation 'pybind11::d
  etail::descr<12>' being compiled
  d:\libs\pybind\2.5.0\src\include\pybind11\cast.h(1602): note: see reference to class template instantiation 'pybind11
  ::detail::descr<8>' being compiled
  d:\libs\pybind\2.5.0\src\include\pybind11\cast.h(1600): note: see reference to class template instantiation 'pybind11
  ::detail::descr<5>' being compiled
  d:\libs\pybind\2.5.0\src\include\pybind11\cast.h(1145): note: see reference to class template instantiation 'pybind11
  ::detail::descr<7>' being compiled
d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(422): error C2893: Failed to specialize function template 'enable_
if<_Test,_Ty>::type pybind11::detail::vector_buffer(Class_ &)' [D:\Libs\Pybind\2.5.0\build_ninja\tests\pybind11_tests.v
cxproj]
          with
          [
              _Ty=void
          ]
  d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(422): note: With the following template arguments:
  d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(422): note: 'Vector=Vector'
  d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(422): note: 'Class_=Class_'
  d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(422): note: 'Args={}'
  test_methods_and_attributes.cpp
  Generating Code...
  Compiling...
  test_modules.cpp
  test_multiple_inheritance.cpp
  test_numpy_array.cpp
  test_numpy_dtypes.cpp
  test_numpy_vectorize.cpp
  test_opaque_types.cpp
  test_operator_overloading.cpp
  test_pickling.cpp
  test_pytypes.cpp
  test_sequences_and_iterators.cpp
  test_smart_ptr.cpp
  test_stl.cpp
  test_stl_binders.cpp
d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(434): error C2672: 'pybind11::detail::vector_buffer': no matching
overloaded function found [D:\Libs\Pybind\2.5.0\build_ninja\tests\pybind11_tests.vcxproj]
  d:\libs\pybind\2.5.0\src\tests\test_stl_binders.cpp(72): note: see reference to function template instantiation 'pybi
  nd11::class_<std::vector<El,std::allocator<_Ty>>,std::unique_ptr<Vector,std::default_delete<std::vector<_Ty,std::allo
  cator<_Ty>>>>> pybind11::bind_vector<std::vector<_Ty,std::allocator<_Ty>>,std::unique_ptr<Vector,std::default_delete<
  std::vector<_Ty,std::allocator<_Ty>>>>,>(pybind11::handle,const std::string &)' being compiled
          with
          [
              _Ty=El,
              Vector=std::vector<El,std::allocator<El>>
          ]
  d:\libs\pybind\2.5.0\src\tests\pybind11_tests.h(60): note: see reference to class template instantiation 'pybind11::d
  etail::descr<12>' being compiled
  d:\libs\pybind\2.5.0\src\include\pybind11\cast.h(1602): note: see reference to class template instantiation 'pybind11
  ::detail::descr<8>' being compiled
  d:\libs\pybind\2.5.0\src\include\pybind11\cast.h(1600): note: see reference to class template instantiation 'pybind11
  ::detail::descr<5>' being compiled
  d:\libs\pybind\2.5.0\src\include\pybind11\cast.h(1145): note: see reference to class template instantiation 'pybind11
  ::detail::descr<7>' being compiled
d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(422): error C2893: Failed to specialize function template 'enable_
if<_Test,_Ty>::type pybind11::detail::vector_buffer(Class_ &)' [D:\Libs\Pybind\2.5.0\build_ninja\tests\pybind11_tests.v
cxproj]
          with
          [
              _Ty=void
          ]
  d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(422): note: With the following template arguments:
  d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(422): note: 'Vector=Vector'
  d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(422): note: 'Class_=Class_'
  d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(422): note: 'Args={}'
  test_tagbased_polymorphic.cpp
  test_union.cpp
  test_virtual_functions.cpp
  Generating Code...
Done Building Project "D:\Libs\Pybind\2.5.0\build_ninja\tests\pybind11_tests.vcxproj" (default targets) -- FAILED.

Done Building Project "D:\Libs\Pybind\2.5.0\build_ninja\ALL_BUILD.vcxproj" (default targets) -- FAILED.


Build FAILED.

"D:\Libs\Pybind\2.5.0\build_ninja\ALL_BUILD.vcxproj" (default target) (1) ->
"D:\Libs\Pybind\2.5.0\build_ninja\tests\pybind11_cross_module_tests.vcxproj" (default target) (4) ->
(ClCompile target) ->
  d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(434): error C2672: 'pybind11::detail::vector_buffer': no matchin
g overloaded function found [D:\Libs\Pybind\2.5.0\build_ninja\tests\pybind11_cross_module_tests.vcxproj]
  d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(422): error C2893: Failed to specialize function template 'enabl
e_if<_Test,_Ty>::type pybind11::detail::vector_buffer(Class_ &)' [D:\Libs\Pybind\2.5.0\build_ninja\tests\pybind11_cross
_module_tests.vcxproj]


"D:\Libs\Pybind\2.5.0\build_ninja\ALL_BUILD.vcxproj" (default target) (1) ->
"D:\Libs\Pybind\2.5.0\build_ninja\tests\pybind11_tests.vcxproj" (default target) (5) ->
  d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(434): error C2672: 'pybind11::detail::vector_buffer': no matchin
g overloaded function found [D:\Libs\Pybind\2.5.0\build_ninja\tests\pybind11_tests.vcxproj]
  d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(422): error C2893: Failed to specialize function template 'enabl
e_if<_Test,_Ty>::type pybind11::detail::vector_buffer(Class_ &)' [D:\Libs\Pybind\2.5.0\build_ninja\tests\pybind11_tests
.vcxproj]
  d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(434): error C2672: 'pybind11::detail::vector_buffer': no matchin
g overloaded function found [D:\Libs\Pybind\2.5.0\build_ninja\tests\pybind11_tests.vcxproj]
  d:\libs\pybind\2.5.0\src\include\pybind11\stl_bind.h(422): error C2893: Failed to specialize function template 'enabl
e_if<_Test,_Ty>::type pybind11::detail::vector_buffer(Class_ &)' [D:\Libs\Pybind\2.5.0\build_ninja\tests\pybind11_tests
.vcxproj]

    0 Warning(s)
    6 Error(s)

Time Elapsed 00:02:27.87
```

I did NOT try with 32 bit compiler and 32 bit python...

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -16,7 +16,6 @@ jobs:
       matrix:
         runs-on: [ubuntu-latest, windows-latest, macos-latest]
         arch: [x64]
-        max-cxx-std: [17]
         python:
         - 2.7
         - 3.5
@@ -28,34 +27,28 @@ jobs:
           - runs-on: ubuntu-latest
             python: 3.6
             arch: x64
-            max-cxx-std: 17
             args: >
               -DPYBIND11_FINDPYTHON=ON
           - runs-on: windows-2016
             python: 3.7
             arch: x86
-            max-cxx-std: 17
             args2: >
               -DCMAKE_CXX_FLAGS="/permissive- /EHsc /GR"
           - runs-on: windows-latest
             python: 3.6
             arch: x64
-            max-cxx-std: 17
             args: >
               -DPYBIND11_FINDPYTHON=ON
           - runs-on: windows-latest
             python: 3.7
             arch: x64
-            max-cxx-std: 17
 
           - runs-on: ubuntu-latest
             python: 3.9-dev
             arch: x64
-            max-cxx-std: 17
           - runs-on: macos-latest
             python: 3.9-dev
             arch: x64
-            max-cxx-std: 17
             args: >
               -DPYBIND11_FINDPYTHON=ON
 
@@ -64,21 +57,17 @@ jobs:
           - runs-on: windows-latest
             python: pypy2
             arch: x64
-            max-cxx-std: 17
           - runs-on: windows-latest
             python: pypy3
             arch: x64
-            max-cxx-std: 17
 
             # Currently broken on embed_test
           - runs-on: windows-latest
             python: 3.8
             arch: x64
-            max-cxx-std: 17
           - runs-on: windows-latest
             python: 3.9-dev
             arch: x64
-            max-cxx-std: 17
 
 
     name: "🐍 ${{ matrix.python }} • ${{ matrix.runs-on }} • ${{ matrix.arch }} ${{ matrix.args }}"
@@ -113,7 +102,7 @@ jobs:
     - name: Prepare env
       run: python -m pip install -r tests/requirements.txt --prefer-binary
 
-    - name: Configure C++11 ${{ matrix.args }}
+    - name: Configure C++11 ${{ matrix.args1 }}
       run: >
         cmake -S . -B .
         -DPYBIND11_WERROR=ON
@@ -137,26 +126,26 @@ jobs:
     - name: Clean directory
       run: git clean -fdx
 
-    - name: Configure C++${{ matrix.max-cxx-std }} ${{ matrix.args }}
+    - name: Configure ${{ matrix.args2 }}
       run: >
         cmake -S . -B build2
         -DPYBIND11_WERROR=ON
         -DDOWNLOAD_CATCH=ON
         -DDOWNLOAD_EIGEN=ON
-        -DCMAKE_CXX_STANDARD=${{ matrix.max-cxx-std }}
+        -DCMAKE_CXX_STANDARD=17
         ${{ matrix.args }}
         ${{ matrix.args2 }}
 
-    - name: Build C++${{ matrix.max-cxx-std }}
+    - name: Build
       run: cmake --build build2 -j 2
 
-    - name: Python tests C++${{ matrix.max-cxx-std }}
+    - name: Python tests
       run: cmake --build build2 --target pytest
 
-    - name: C++${{ matrix.max-cxx-std }} tests
+    - name: C++ tests
       run: cmake --build build2 --target cpptest
 
-    - name: Interface test C++${{ matrix.max-cxx-std }}
+    - name: Interface test
       run: cmake --build build2 --target test_cmake_build
 
   clang:

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -29,21 +29,20 @@ jobs:
             python: 3.6
             arch: x64
             max-cxx-std: 17
-            args: "-DPYBIND11_FINDPYTHON=ON"
-          - runs-on: macos-latest
-            python: 3.7
-            arch: x64
-            max-cxx-std: 17
-            args: "-DPYBIND11_FINDPYTHON=ON"
+            args: >
+              -DPYBIND11_FINDPYTHON=ON
           - runs-on: windows-2016
             python: 3.7
             arch: x86
-            max-cxx-std: 14
+            max-cxx-std: 17
+            args2: >
+              -DCMAKE_CXX_FLAGS="/permissive- /EHsc /GR"
           - runs-on: windows-latest
             python: 3.6
             arch: x64
             max-cxx-std: 17
-            args: "-DPYBIND11_FINDPYTHON=ON"
+            args: >
+              -DPYBIND11_FINDPYTHON=ON
           - runs-on: windows-latest
             python: 3.7
             arch: x64
@@ -57,6 +56,8 @@ jobs:
             python: 3.9-dev
             arch: x64
             max-cxx-std: 17
+            args: >
+              -DPYBIND11_FINDPYTHON=ON
 
         exclude:
             # Currently 32bit only, and we build 64bit
@@ -92,8 +93,7 @@ jobs:
         python-version: ${{ matrix.python }}
         architecture: ${{ matrix.arch }}
 
-    - name: Setup Boost
-      if: runner.os != 'macOS'
+    - name: Setup Boost (Windows / Linux latest)
       run: echo "::set-env name=BOOST_ROOT::$BOOST_ROOT_1_72_0"
 
     - name: Update CMake
@@ -114,7 +114,6 @@ jobs:
       run: python -m pip install -r tests/requirements.txt --prefer-binary
 
     - name: Configure C++11 ${{ matrix.args }}
-      shell: bash
       run: >
         cmake -S . -B .
         -DPYBIND11_WERROR=ON
@@ -139,7 +138,6 @@ jobs:
       run: git clean -fdx
 
     - name: Configure C++${{ matrix.max-cxx-std }} ${{ matrix.args }}
-      shell: bash
       run: >
         cmake -S . -B build2
         -DPYBIND11_WERROR=ON
@@ -147,6 +145,7 @@ jobs:
         -DDOWNLOAD_EIGEN=ON
         -DCMAKE_CXX_STANDARD=${{ matrix.max-cxx-std }}
         ${{ matrix.args }}
+        ${{ matrix.args2 }}
 
     - name: Build C++${{ matrix.max-cxx-std }}
       run: cmake --build build2 -j 2

--- a/docs/basics.rst
+++ b/docs/basics.rst
@@ -35,6 +35,14 @@ Windows
 On Windows, only **Visual Studio 2015** and newer are supported since pybind11 relies
 on various C++11 language features that break older versions of Visual Studio.
 
+.. Note::
+
+    To use the C++17 in Visual Studio 2017 (MSVC 14.1), pybind11 requires the flag
+    ``/permissive-`` to be passed to the compiler `to enforce standard conformance`_. When
+    building with Visual Studio 2019, this is not strictly necessary, but still adviced.
+
+..  _`to enforce standard conformance`: https://docs.microsoft.com/en-us/cpp/build/reference/permissive-standards-conformance?view=vs-2017
+
 To compile and run the tests:
 
 .. code-block:: batch

